### PR TITLE
Unit tests cleanup for sstable generation changes

### PR DESCRIPTION
**Pull request by bhalevy**
This series cleans up unit test in preparation for PR #12994.
Helpers are added (or reused) to not rely on specific sstable generation numbers where possible (other than loading reference sstables that are committed to the repo with given generation numbers), and to generate the sstables for tests easily, taking advantage of generation management in `sstable_test_env`, `table_for_tests`, or `replica::table` itself.

### CODE REVIEW

File: test/boost/sstable_compaction_test.cc [hunk 1]
```diff
@@ -211,21 +211,25 @@ SEASTAR_TEST_CASE(compact) {
...
                 //   john |  20 |   deleted
                 //   nadav - deleted partition
-                return env.reusable_sst(s, generation).then([&env, s] (shared_sstable sst) {
-                    auto reader = make_lw_shared<flat_mutation_reader_v2>(sstable_reader(sst, s, env.make_reader_permit())); // reader holds sst and s alive.
-                    return read_mutation_from_flat_mutation_reader(*reader).then([reader, s] (mutation_opt m) {
+                auto sst = env.reusable_sst(s, generation).get();
+                    auto reader = sstable_reader(sst, s, env.make_reader_permit());
+                    auto close_reader = deferred_close(reader);
+                    auto verify_mutation = [&] (std::function<void(mutation_opt)> verify) {
+                        std::invoke(verify, read_mutation_from_flat_mutation_reader(reader).get());
```
Comment (avikivity) [inline test/boost/sstable_compaction_test.cc]: Why std::invoke? You can write verify(read_mutation_from_flat_mutation_reader().get());. But don't bother changing.

File: test/boost/sstable_datafile_test.cc [hunk 2]
```diff
@@ -181,47 +179,47 @@ SEASTAR_TEST_CASE(datafile_generation_11) {
...
-                        scol->as_collection_mutation().with_deserialized(*static_set_col.type, [&] (collection_mutation_view_description mut) {
-                            verify_set(mut.materialize(*static_set_col.type));
-                        });
-
-                        // The clustered set
-                        auto m = verifier(mutation);
-                        verify_set(m);
-            });
+        auto sstp = env.reusable_sst(s, 11).get();
+        std::invoke([&] {
```
Comment (avikivity) [inline test/boost/sstable_datafile_test.cc]: What's the point of std::invoke here?

File: test/boost/sstable_3_x_test.cc [hunk 3]
```diff
@@ -3035,8 +3035,7 @@ static flat_mutation_reader_v2 compacted_sstable_reader(test_env& env, schema_pt
 
     auto desc = sstables::compaction_descriptor(std::move(sstables), default_priority_class());
     desc.creator = [&] (shard_id dummy) {
-        compacted_sst = env.make_sstable(s, tmp.path().string(), new_generation,
-                         sstables::sstable_version_types::mc, sstable::format_types::big, 4096);
```
Comment (xemul) [inline test/boost/sstable_3_x_test.cc:3033]: This defaults the buffer-size of 4096. Dunno if this is was test's intention to use smaller buffer

File: test/boost/mutation_reader_test.cc [hunk 4]
```diff
@@ -1000,10 +995,9 @@ SEASTAR_TEST_CASE(reader_selector_fast_forwarding_test) {
 
 static
 sstables::shared_sstable create_sstable(sstables::test_env& env, schema_ptr s, std::vector<mutation> mutations) {
-    static thread_local auto tmp = tmpdir();
     static int gen = 0;
     return make_sstable_containing([&] {
-        return env.make_sstable(s, tmp.path().string(), gen++);
+        return env.make_sstable(s, gen++);
```
Comment (xemul) [inline test/boost/mutation_reader_test.cc]: nit: You can drop gen++ too, sstables::test_env maintains monotonically incrementing generation on its own

File: test/boost/sstable_datafile_test.cc [hunk 5]
```diff
@@ -470,7 +470,7 @@ SEASTAR_TEST_CASE(datafile_generation_47) {
 
         auto sst = env.make_sstable(s, 47);
         write_memtable_to_sstable_for_test(*mt, sst).get();
-        auto sstp = env.reusable_sst(s, 47).get();
+        auto sstp = env.reusable_sst(s, sst).get();
```
Comment (xemul) [inline test/boost/sstable_datafile_test.cc]: The `s` seem to be the same as for `sst`, why not call schema-less reusable_sst() overload?

File: test/lib/sstable_test_env.hh [hunk 6]
```diff
@@ -127,6 +127,14 @@ public:
         return reusable_sst(std::move(schema), _impl->dir.path().native(), generation);
     }
 
+    future<shared_sstable> reusable_sst(schema_ptr schema, shared_sstable sst) {
```
Comment (xemul) [inline test/lib/sstable_test_env.hh:160]: I've came across a call to this overload above, but then noticed that there are many other places that call it and it wasn't obvious why. So my question is -- if the intention of this overload is to re-open an sstable from an existing one, what could be the reason for _not_ using the `sst`'s schema?

File: test/lib/sstable_utils.cc [hunk 7]
```diff
@@ -208,3 +208,35 @@ void compaction_manager_test::deregister_compaction(const sstables::compaction_d
         testlog.error("compaction_manager_test: deregister_compaction uuid={}: task not found", c.compaction_uuid);
     }
 }
+
+shared_sstable verify_mutation(test_env& env, shared_sstable sst, lw_shared_ptr<replica::memtable> mt, bytes key, std::function<void(mutation_opt&)> verify) {
+    auto sstp = make_sstable_containing(std::move(sst), mt);
+    return verify_mutation(env, std::move(sstp), std::move(key), std::move(verify));
+}
+
+shared_sstable verify_mutation(test_env& env, shared_sstable sstp, bytes key, std::function<void(mutation_opt&)> verify) {
```
Comment (xemul) [inline test/lib/sstable_utils.cc:217]: There was `verify_mutation` lambda in this PR (in sstable_compaction_test). They look very much alike and deserve being merged/generalized

File: test/boost/sstable_3_x_test.cc [hunk 8]
```diff
@@ -3216,41 +3220,39 @@ static void compare_sstables(const std::filesystem::path& result_path, sstring t
...
 
-static tmpdir write_sstables(test_env& env, schema_ptr s, lw_shared_ptr<replica::memtable> mt1, lw_shared_ptr<replica::memtable> mt2, sstable_version_types version) {
-    tmpdir tmp;
-    auto sst = env.make_sstable(s, tmp.path().string(), 1, version, sstable::format_types::big, 4096);
+static void compare_sstables(test_env& env, sstring table_name, sstable_version_types version) {
+    compare_sstables(env.tempdir().path().native(), std::move(table_name), version);
+}
+
+static void write_sstables(test_env& env, schema_ptr s, lw_shared_ptr<replica::memtable> mt1, lw_shared_ptr<replica::memtable> mt2, sstable_version_types version) {
+    auto sst = env.make_sstable(s, env.tempdir().path().string(), 1, version, sstable::format_types::big, 4096);
```
Comment (xemul) [inline test/boost/sstable_3_x_test.cc]: sstables::test_env() has a dir-less overload that uses its tempdir transparently

File: test/boost/sstable_compaction_test.cc [hunk 9]
```diff
@@ -3851,7 +3852,7 @@ SEASTAR_TEST_CASE(test_offstrategy_sstable_compaction) {
             auto mut = mutation(s, pk);
             ss.add_row(mut, ss.make_ckey(0), "val");
 
-            table_for_tests cf(env.manager(), s, tmp.path().string());
```
Comment (xemul) [inline test/boost/sstable_compaction_test.cc:3897]: It looks like tmp is the back() element of a vector which doesn't have to be a vector -- test just creates new tempdir on next loop iteration and doesn't mess with previous one.

File: test/lib/sstable_test_env.hh [hunk 10]
```diff
@@ -185,6 +185,10 @@ public:
             return func(env);
         });
     }
+
+    table_for_tests make_table_for_tests(schema_ptr s = nullptr, std::optional<sstring> dir = std::nullopt) {
```
Comment (xemul) [inline test/lib/sstable_test_env.hh]: Taking into account my comment above, the dir argument is not used and can be removed. This will discourage users from using explicit tempdir which's good

File: test/boost/sstable_3_x_test.cc [hunk 11]
```diff
@@ -3028,7 +3027,6 @@ static flat_mutation_reader_v2 compacted_sstable_reader(test_env& env, schema_pt
     cf->mark_ready_for_writes();
     lw_shared_ptr<replica::memtable> mt = make_lw_shared<replica::memtable>(s);
 
-    tmpdir tmp;
```
Comment (xemul) [inline test/boost/sstable_3_x_test.cc:3026]: nit: This belongs to separate patch "drop unused tempdir"

File: test/boost/sstable_compaction_test.cc [hunk 12]
```diff
@@ -3438,7 +3438,6 @@ SEASTAR_TEST_CASE(autocompaction_control_test) {
                 .with_column("value", int32_type)
                 .build();
 
-        auto tmp = tmpdir();
```
Comment (xemul) [inline test/boost/sstable_compaction_test.cc]: nit: This also belongs to separate patch "drop unused tempdir"

File: test/boost/sstable_datafile_test.cc [hunk 13]
```diff
@@ -1151,7 +1152,6 @@ SEASTAR_TEST_CASE(sstable_tombstone_metadata_check) {
                     .with_column("ck1", utf8_type, column_kind::clustering_key)
                     .with_column("r1", int32_type)
                     .build();
-            auto tmp = env.tempdir().make_sweeper();
```
Comment (xemul) [inline test/boost/sstable_datafile_test.cc:1206]: But why? I placed sweepers in locations where test cases failed stepping on non-empty directory. Why did it become useless?

File: test/boost/sstable_datafile_test.cc [hunk 14]
```diff
@@ -3169,7 +3168,7 @@ SEASTAR_TEST_CASE(find_first_position_in_partition_from_sstable_test) {
         class with_static_row_tag;
         using with_static_row = bool_class<with_static_row_tag>;
 
-        auto check_sstable_first_and_last_positions = [&env] (size_t partitions, with_range_tombstone with_range_tombstone, with_static_row with_static_row) {
+        auto check_sstable_first_and_last_positions = [&] (size_t partitions, with_range_tombstone with_range_tombstone, with_static_row with_static_row) {
```
Comment (xemul) [inline test/boost/sstable_datafile_test.cc:3058]: Unrelated?

File: test/boost/sstable_directory_test.cc [hunk 15]
```diff
@@ -111,21 +123,36 @@ highest_generation_seen(sharded<sstables::sstable_directory>& dir) {
 
 class wrapped_test_env {
```
Comment (xemul) [inline test/boost/sstable_directory_test.cc:124]: The wrapped_test_env was supposed to provide a way to extract sstables_manager from a test_env whatever the env is. Now it's turning into something bigger which deserves dedicated explanation

File: test/boost/sstable_directory_test.cc [hunk 16]
```diff
@@ -95,7 +96,18 @@ make_sstable_for_all_shards(replica::database& db, replica::table& table, fs::pa
...
 
 sstables::shared_sstable new_sstable(sstables::test_env& env, fs::path dir, int64_t gen) {
-    return env.make_sstable(test_table_schema(), dir.native(), generation_from_value(gen));
+    testlog.debug("new_sstable: dir={} gen={}", dir, gen);
+    return env.make_sstable(test_table_schema(), dir.native(), gen);
+}
+
+sstables::shared_sstable new_env_sstable(sstables::test_env& env) {
+    testlog.debug("new_env_sstable: dir={}", env.tempdir().path());
+    return env.make_sstable(test_table_schema());
```
Comment (xemul) [inline test/boost/sstable_directory_test.cc:105]: The existing new_sstable() calls env.manager().make_sstable()
The new new_env_sstable() calls env.make_sstable() which, in turn, calls env.manager().make_sstable()

Is this helper really needed?

**Comment by nyh** [comment]
@bhalevy it seems that @avikivity merged your series before @xemul submitted a review. Please consider what @xemul commented and consider if you want to send a followup.

**Comment by bhalevy** [comment]
> @bhalevy it seems that @avikivity merged your series before @xemul submitted a review. Please consider what @xemul commented and consider if you want to send a followup.

of course

File: sstables/sstables.hh [hunk 17]
```diff
@@ -507,6 +507,10 @@ public:
         sstring prefix() const { return dir; }
     };
 
+    const filesystem_storage& get_storage() const {
```
Comment (xemul) [inline sstables/sstables.hh:510]: It's better not to, and since it's used only by tests, the `class sstables::test` is the way to put this method into
